### ISSUE TITLE
Loss doesn't converge when training on my own data

### ISSUE THREAD

**Issue by garryz94** [HuguesTHOMAS/KPConv#77]
Hi Hugues, thanks for your excellent work! 
I have been training KPConv segmentation model on my own data recently. In the training process, the train loss decreases at the beginning. But after a few epochs(about 20 epochs), the train loss as well as the validation loss doesn't decrease any more. 
As I am facing a severe problem of imbalanced data, I replaced the ce loss with multi-class focal loss. But the train loss doesn't decrease either. At first, I thought I was encountering with an underfitting problem,  so I tried deeper and wider network, but it doesn't help.
So can you tell me what happened in the network and what should I do to improve the segmentation performance? Thanks again for your brilliant work, hope to get your reply soon!

**Comment by HuguesTHOMAS** [comment]
Hi @garryz94,

I cant know for sure without seeing your dataset, but there are a few tricks that could help when dealing with imbalanced data. I already gave a detailed answer for the PyTorch version:
https://github.com/HuguesTHOMAS/KPConv-PyTorch/issues/2#issuecomment-626333820

The answer is overall the same for Tensorflow version. Try to use the balanced class data loader. In Tensorflow you have to modify the code directly in the dataset class. For example in S3DIS:
https://github.com/HuguesTHOMAS/KPConv/blob/132fdc628fb4850548e931c8b02c6325e7cac85e/datasets/S3DIS.py#L641-L649

Use the `random_balanced_gen` defined line 444:
https://github.com/HuguesTHOMAS/KPConv/blob/132fdc628fb4850548e931c8b02c6325e7cac85e/datasets/S3DIS.py#L444

Such a dataloader with a smaller input radius will help the network to "see" more of the minority classes.

Concerning the loss, I did not implement the class weights in Tensorflow, but you already tried your own multi-class focal loss, so you can still use that, or figure out another weighted class loss. 

Good luck with your problem.
Best,
Hugues

**Comment by garryz94** [comment]
> Hi @garryz94,
> 
> I cant know for sure without seeing your dataset, but there are a few tricks that could help when dealing with imbalanced data. I already gave a detailed answer for the PyTorch version:
> [HuguesTHOMAS/KPConv-PyTorch#2 (comment)](https://github.com/HuguesTHOMAS/KPConv-PyTorch/issues/2#issuecomment-626333820)
> 
> The answer is overall the same for Tensorflow version. Try to use the balanced class data loader. In Tensorflow you have to modify the code directly in the dataset class. For example in S3DIS:
> https://github.com/HuguesTHOMAS/KPConv/blob/132fdc628fb4850548e931c8b02c6325e7cac85e/datasets/S3DIS.py#L641-L649
> 
> Use the `random_balanced_gen` defined line 444:
> https://github.com/HuguesTHOMAS/KPConv/blob/132fdc628fb4850548e931c8b02c6325e7cac85e/datasets/S3DIS.py#L444
> 
> Such a dataloader with a smaller input radius will help the network to "see" more of the minority classes.
> 
> Concerning the loss, I did not implement the class weights in Tensorflow, but you already tried your own multi-class focal loss, so you can still use that, or figure out another weighted class loss.
> 
> Good luck with your problem.
> Best,
> Hugues

Thanks for your quick reply, I will try the data loader!
By the way, I am now working on CURB recognition. In general, curb is "strip-type" and has fewer points compared to other classes like ground, building, etc... I tried to adjust the shape of kernel points(from sphere to ellipsoid) in order to adapt to “strip-type” object segmentation. But I still can't get a better performance. Can you give me some advice on recognizing "strip-type" object？Thanks！

**Comment by HuguesTHOMAS** [comment]
Changing the kernel shape wont help in my opinion. The reason could be that the shape of a curb is two small for the network to "see" it. You might want to reduce the `first_subsampling_dl` so that the first kernel size is smaller than the size of a curb. Something like 2 cm. You will have to reduce the input radius accordingly, so if you also have to detect bigger object, maybe you should use two different networks.

**Comment by garryz94** [comment]
> Changing the kernel shape wont help in my opinion. The reason could be that the shape of a curb is two small for the network to "see" it. You might want to reduce the `first_subsampling_dl` so that the first kernel size is smaller than the size of a curb. Something like 2 cm. You will have to reduce the input radius accordingly, so if you also have to detect bigger object, maybe you should use two different networks.

Thanks! I used **random_balanced_gen** in training, it really helps when combined with focal loss. But **class_weight** seems to be less useful or even harmful when using **random_balanced_gen**. I think the reason is that the data distribution generated by **random_balanced_gen** is quite different from the original distribution.
As for CURB detection, your opinion about kernel shape is right. As curb is the only class that is important to me, I might modify the **random_balanced_gen** a little bit to generate more data of curb. So the model could "see" more samples of curb and I hope the recognition ability of the only class will benefit from it. I think raise the class_weight of curb a little bit can help too.'
Thanks again for your excellent work!

**Comment by HuguesTHOMAS** [comment]
> I think the reason is that the data distribution generated by random_balanced_gen is quite different from the original distribution.

You are right. If you want to use class weights, you have to create a small function that count the points points from each class in the input sphere. This will give you the real class distribution that the network sees.

> As for CURB detection, your opinion about kernel shape is right. As curb is the only class that is important to me, I might modify the random_balanced_gen a little bit to generate more data of curb. So the model could "see" more samples of curb and I hope the recognition ability of the only class will benefit from it. I think raise the class_weight of curb a little bit can help too.'
Thanks again for your excellent work!

Good luck with your problem.

**Comment by garryz94** [comment]
Thanks for your help!
